mobile_config/src/hotspot_service.rs

```diff
@@ -0,0 +1,178 @@
+            .map_or_else(
+                || Err(Status::not_found(pubkey.to_string())),
+                |metadata| {
+                    let metadata = metadata
+                        .try_into()
+                        .map_err(|_| Status::internal("error serializing hotspot metadata"))?;
+                    let mut res = HotspotMetadataResV1 {
+                        metadata: Some(metadata),
+                        timestamp: Utc::now().encode_timestamp(),
+                        signature: vec![],
+                    };
+                    res.signature = self.sign_response(&res)?;
+                    Ok(Response::new(res))
+                },
+            )
+    }
+
+    type metadata_streamStream = GrpcStreamResult<HotspotMetadataStreamResV1>;
+    async fn metadata_stream(
+        &self,
+        request: Request<HotspotMetadataStreamReqV1>,
+    ) -> GrpcResult<Self::metadata_streamStream> {
+        let request = request.into_inner();
+
+        self.verify_request_signature(&request).await?;
+
+        tracing::debug!("fetching all hotspots' metadata");
+
+        let pool = self.metadata_pool.clone();
+        let signing_key = self.signing_key.clone();
+        let batch_size = request.batch_size;
+
+        let (tx, rx) = tokio::sync::mpsc::channel(20);
+
+        tokio::spawn(async move {
+            stream_all_hotspots_metadata(&pool, tx.clone(), signing_key.clone(), batch_size).await
+        });
+
+        Ok(Response::new(GrpcStreamResult::new(rx)))
+    }
+}
+
+async fn stream_all_hotspots_metadata(
+    pool: &Pool<Postgres>,
+    tx: tokio::sync::mpsc::Sender<Result<HotspotMetadataStreamResV1, Status>>,
+    signing_key: Arc<Keypair>,
+    batch_size: u32,
+) -> anyhow::Result<()> {
+    Ok(hotspot_metadata::all_metadata_stream(pool)
+        .map(Ok::<HotspotMetadata, sqlx::Error>)
+        .try_filter_map(|metadata| async move {
+            let result: Option<mobile_config::HotspotMetadata> = match metadata.try_into() {
+                Ok(metadata_proto) => Some(metadata_proto),
+                Err(_) => None,
+            };
+            Ok(result)
+        })
+        .try_chunks(batch_size as usize)
+        .map_ok(move |batch| {
+            (
+                HotspotMetadataStreamResV1 {
+                    hotspots: batch,
+                    timestamp: Utc::now().encode_timestamp(),
+                    signature: vec![],
+                },
+                signing_key.clone(),
+            )
+        })
+        .try_filter_map(|(res, keypair)| async move {
+            let result = match keypair.sign(&res.encode_to_vec()) {
+                Ok(signature) => Some(HotspotMetadataStreamResV1 {
+                    hotspots: res.hotspots,
+                    timestamp: res.timestamp,
+                    signature,
+                }),
+                Err(_) => None,
+            };
+            Ok(result)
+        })
+        .or_else(|err| async move {
```

Comment:
You should be able to use a `map_err` here which doesn't need to return a Future so it should increase readability a little

Reply:
[eb412d3](https://github.com/helium/oracles/pull/419/commits/eb412d3d6af761ab3ec6bae92ef61e4f7792f01e) i swapped the first `or_else` out for a `map_err` but then in the followup i reworked the final error map into an `or_else` to catch the errors and resend them on to the client to decide how to handle them (or not) since the stream is technically a stream of `Result<HotspotInfo, Status>`